// @livekit/react-native/lib/commonjs/components/VideoTrack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VideoTrack = void 0;
var _reactNative = require("react-native");
var _livekitClient = require("livekit-client");
var _reactNativeWebrtc = require("@livekit/react-native-webrtc");
var _react = require("react");
var _ViewPortDetector = _interopRequireDefault(require("./ViewPortDetector"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Props for the VideoTrack component.
 * @public
 */

/**
 * @inline
 */

/**
 * VideoTrack component for displaying video tracks in a React Native application.
 * It supports both local and remote video tracks from LiveKit, and handles adaptive streaming for remote tracks.
 *
 * @param props - See VideoTrackProps for details.
 * @returns A React component that renders the given video track.
 * @public
 */
const VideoTrack = exports.VideoTrack = /*#__PURE__*/(0, _react.forwardRef)(({
  style = {},
  trackRef,
  objectFit = 'cover',
  zOrder,
  mirror,
  iosPIP
}, ref) => {
  const [elementInfo] = (0, _react.useState)(() => {
    let info = new VideoTrackElementInfo();
    info.id = trackRef?.publication?.trackSid;
    return info;
  });
  const layoutOnChange = (0, _react.useCallback)(event => elementInfo.onLayout(event), [elementInfo]);
  const iosPIPEnabled = iosPIP?.enabled ?? false;
  const visibilityOnChange = (0, _react.useCallback)(isVisible => elementInfo.onVisibility(isVisible || iosPIPEnabled), [elementInfo, iosPIPEnabled]);
  const videoTrack = trackRef?.publication.track;
  const shouldObserveVisibility = (0, _react.useMemo)(() => {
    return videoTrack instanceof _livekitClient.RemoteVideoTrack && videoTrack.isAdaptiveStream;
  }, [videoTrack]);
  const [mediaStream, setMediaStream] = (0, _react.useState)(videoTrack?.mediaStream);
  (0, _react.useEffect)(() => {
    setMediaStream(videoTrack?.mediaStream);
    if (videoTrack instanceof _livekitClient.LocalVideoTrack) {
      const onRestarted = track => {
        setMediaStream(track?.mediaStream);
      };
      videoTrack.on(_livekitClient.TrackEvent.Restarted, onRestarted);
      return () => {
        videoTrack.off(_livekitClient.TrackEvent.Restarted, onRestarted);
      };
    } else {
      return () => {};
    }
  }, [videoTrack]);
  (0, _react.useEffect)(() => {
    if (videoTrack instanceof _livekitClient.RemoteVideoTrack && videoTrack.isAdaptiveStream) {
      videoTrack?.observeElementInfo(elementInfo);
      return () => {
        videoTrack?.stopObservingElementInfo(elementInfo);
      };
    } else {
      return () => {};
    }
  }, [videoTrack, elementInfo]);
  let videoView;
  if (!iosPIP || _reactNative.Platform.OS !== 'ios') {
    videoView = /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNativeWebrtc.RTCView, {
      style: styles.videoTrack
      // @ts-expect-error
      ,
      streamURL: mediaStream?.toURL() ?? '',
      objectFit: objectFit,
      zOrder: zOrder,
      mirror: mirror
      // @ts-ignore
      ,
      ref: ref
    });
  } else {
    videoView = /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNativeWebrtc.RTCPIPView, {
      style: styles.videoTrack
      // @ts-expect-error
      ,
      streamURL: mediaStream?.toURL() ?? '',
      objectFit: objectFit,
      zOrder: zOrder,
      mirror: mirror
      // TODO: fix this up in react-native-webrtc side.
      // @ts-expect-error
      ,
      iosPIP: iosPIP,
      ref: ref
    });
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {
    style: {
      ...style,
      ...styles.container
    },
    onLayout: layoutOnChange,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_ViewPortDetector.default, {
      onChange: visibilityOnChange,
      style: styles.videoTrack,
      disabled: !shouldObserveVisibility,
      propKey: videoTrack,
      children: videoView
    })
  });
});
const styles = _reactNative.StyleSheet.create({
  container: {},
  videoTrack: {
    flex: 1,
    width: '100%'
  }
});
class VideoTrackElementInfo {
  element = {};
  _width = 0;
  _height = 0;
  _observing = false;
  visible = true;
  pictureInPicture = false;
  width = () => this._width;
  height = () => this._height;
  observe() {
    this._observing = true;
  }
  stopObserving() {
    this._observing = false;
  }
  onLayout(event) {
    let {
      width,
      height
    } = event.nativeEvent.layout;
    this._width = width;
    this._height = height;
    if (this._observing) {
      this.handleResize?.();
    }
  }
  onVisibility(isVisible) {
    if (this.visible !== isVisible) {
      this.visible = isVisible;
      this.visibilityChangedAt = Date.now();
      if (this._observing) {
        this.handleVisibilityChanged?.();
      }
    }
  }
}
