// @livekit/react-native/lib/commonjs/components/VideoView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VideoView = void 0;
var _reactNative = require("react-native");
var _livekitClient = require("livekit-client");
var _reactNativeWebrtc = require("@livekit/react-native-webrtc");
var _react = require("react");
var _ViewPortDetector = _interopRequireDefault(require("./ViewPortDetector"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * @deprecated use `VideoTrack` and `VideoTrackProps` instead.
 */

/**
 * @deprecated use `VideoTrack` and `VideoTrackProps` instead.
 */
const VideoView = ({
  style = {},
  videoTrack,
  objectFit = 'cover',
  zOrder,
  mirror
}) => {
  const [elementInfo] = (0, _react.useState)(() => {
    let info = new VideoViewElementInfo();
    info.id = videoTrack?.sid;
    info.something = videoTrack;
    return info;
  });
  const layoutOnChange = (0, _react.useCallback)(event => elementInfo.onLayout(event), [elementInfo]);
  const visibilityOnChange = (0, _react.useCallback)(isVisible => elementInfo.onVisibility(isVisible), [elementInfo]);
  const shouldObserveVisibility = (0, _react.useMemo)(() => {
    return videoTrack instanceof _livekitClient.RemoteVideoTrack && videoTrack.isAdaptiveStream;
  }, [videoTrack]);
  const [mediaStream, setMediaStream] = (0, _react.useState)(videoTrack?.mediaStream);
  (0, _react.useEffect)(() => {
    setMediaStream(videoTrack?.mediaStream);
    if (videoTrack instanceof _livekitClient.LocalVideoTrack) {
      const onRestarted = track => {
        setMediaStream(track?.mediaStream);
      };
      videoTrack.on(_livekitClient.TrackEvent.Restarted, onRestarted);
      return () => {
        videoTrack.off(_livekitClient.TrackEvent.Restarted, onRestarted);
      };
    } else {
      return () => {};
    }
  }, [videoTrack]);
  (0, _react.useEffect)(() => {
    if (videoTrack instanceof _livekitClient.RemoteVideoTrack && videoTrack.isAdaptiveStream) {
      videoTrack?.observeElementInfo(elementInfo);
      return () => {
        videoTrack?.stopObservingElementInfo(elementInfo);
      };
    } else {
      return () => {};
    }
  }, [videoTrack, elementInfo]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {
    style: {
      ...style,
      ...styles.container
    },
    onLayout: layoutOnChange,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_ViewPortDetector.default, {
      onChange: visibilityOnChange,
      style: styles.videoView,
      disabled: !shouldObserveVisibility,
      propKey: videoTrack,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNativeWebrtc.RTCView, {
        style: styles.videoView
        // @ts-expect-error
        ,
        streamURL: mediaStream?.toURL() ?? '',
        objectFit: objectFit,
        zOrder: zOrder,
        mirror: mirror
      })
    })
  });
};
exports.VideoView = VideoView;
const styles = _reactNative.StyleSheet.create({
  container: {},
  videoView: {
    flex: 1,
    width: '100%'
  }
});
class VideoViewElementInfo {
  element = {};
  _width = 0;
  _height = 0;
  _observing = false;
  visible = true;
  pictureInPicture = false;
  width = () => this._width;
  height = () => this._height;
  observe() {
    this._observing = true;
  }
  stopObserving() {
    this._observing = false;
  }
  onLayout(event) {
    let {
      width,
      height
    } = event.nativeEvent.layout;
    this._width = width;
    this._height = height;
    if (this._observing) {
      this.handleResize?.();
    }
  }
  onVisibility(isVisible) {
    if (this.visible !== isVisible) {
      this.visible = isVisible;
      this.visibilityChangedAt = Date.now();
      if (this._observing) {
        this.handleVisibilityChanged?.();
      }
    }
  }
}
