// @livekit/react-native/lib/commonjs/audio/AudioManager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setupIOSAudioManagement = setupIOSAudioManagement;
var _reactNative = require("react-native");
var _AudioSession = _interopRequireDefault(require("./AudioSession"));
var _logger = require("../logger");
var _reactNativeWebrtc = require("@livekit/react-native-webrtc");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const kAudioEngineErrorFailedToConfigureAudioSession = -4100;
let activeAudioManagementSetup;

/**
 * @inline
 */

/**
 * Sets up automatic iOS audio session management based on audio engine state.
 *
 * Call this once at app startup (e.g. in index.js). `registerGlobals()`
 * invokes it for you by default unless `autoConfigureAudioSession: false`
 * is passed.
 *
 * @param preferSpeakerOutput - Whether to prefer speaker output. Defaults to true.
 * @param onConfigureNativeAudio - Optional custom callback for determining audio configuration.
 * @returns A cleanup function that removes the event handlers.
 */
function setupIOSAudioManagement(preferSpeakerOutput = true, onConfigureNativeAudio) {
  if (_reactNative.Platform.OS !== 'ios') {
    return () => {};
  }
  const setupToken = {};
  activeAudioManagementSetup = setupToken;
  let audioEngineState = {
    isPlayoutEnabled: false,
    isRecordingEnabled: false,
    preferSpeakerOutput
  };
  const tryConfigure = async (newState, oldState) => {
    if (!newState.isPlayoutEnabled && !newState.isRecordingEnabled && (oldState.isPlayoutEnabled || oldState.isRecordingEnabled)) {
      _logger.log.info('AudioSession deactivating...');
      await _AudioSession.default.stopAudioSession();
    } else if (newState.isRecordingEnabled || newState.isPlayoutEnabled) {
      const config = onConfigureNativeAudio ? onConfigureNativeAudio(newState) : getDefaultAppleAudioConfigurationForAudioState(newState);
      _logger.log.info('AudioSession configuring category:', config.audioCategory);
      await _AudioSession.default.setAppleAudioConfiguration(config);
      if (!oldState.isPlayoutEnabled && !oldState.isRecordingEnabled) {
        _logger.log.info('AudioSession activating...');
        await _AudioSession.default.startAudioSession();
      }
    }
  };
  const handleEngineStateUpdate = async ({
    isPlayoutEnabled,
    isRecordingEnabled
  }) => {
    const oldState = audioEngineState;
    const newState = {
      isPlayoutEnabled,
      isRecordingEnabled,
      preferSpeakerOutput: audioEngineState.preferSpeakerOutput
    };

    // If tryConfigure throws, the error propagates to the native audio engine
    // observer which converts it to a non-zero error code, causing the engine
    // to stop/rollback (matching the Swift SDK's error propagation pattern).
    try {
      await tryConfigure(newState, oldState);
    } catch (error) {
      _logger.log.error('AudioSession configuration failed, stopping audio engine:', error);
      // Throw the error code so the native AudioDeviceModuleObserver returns it
      // to the WebRTC engine, which will stop/rollback the operation.

      throw kAudioEngineErrorFailedToConfigureAudioSession;
    }
    // Update the audio state only if configure succeeds
    audioEngineState = newState;
  };
  _reactNativeWebrtc.audioDeviceModuleEvents.setWillEnableEngineHandler(handleEngineStateUpdate);
  _reactNativeWebrtc.audioDeviceModuleEvents.setDidDisableEngineHandler(handleEngineStateUpdate);
  return () => {
    if (activeAudioManagementSetup !== setupToken) {
      return;
    }
    activeAudioManagementSetup = undefined;
    _reactNativeWebrtc.audioDeviceModuleEvents.setWillEnableEngineHandler(null);
    _reactNativeWebrtc.audioDeviceModuleEvents.setDidDisableEngineHandler(null);
  };
}

// Kept in sync with `getDefaultAppleAudioConfigurationForMode` in
// `./AudioManagerLegacy.ts`. If you change the defaults in one place,
// update the other so the legacy path and the new path stay aligned.
function getDefaultAppleAudioConfigurationForAudioState(configurationState) {
  if (configurationState.isRecordingEnabled) {
    return {
      audioCategory: 'playAndRecord',
      audioCategoryOptions: ['allowBluetooth', 'mixWithOthers'],
      audioMode: configurationState.preferSpeakerOutput ? 'videoChat' : 'voiceChat'
    };
  } else if (configurationState.isPlayoutEnabled) {
    return {
      audioCategory: 'playback',
      audioCategoryOptions: ['mixWithOthers'],
      audioMode: 'spokenAudio'
    };
  }
  return {
    audioCategory: 'soloAmbient',
    audioCategoryOptions: [],
    audioMode: 'default'
  };
}
