// @livekit/react-native/lib/commonjs/audio/MediaRecorder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MediaRecorder = void 0;
var _reactNativeWebrtc = require("@livekit/react-native-webrtc");
var _EventEmitter = require("../events/EventEmitter");
var _base64Js = require("base64-js");
var _LKNativeModule = _interopRequireDefault(require("../LKNativeModule"));
var _logger = require("../logger");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// typeof MediaRecorder
// const Tester = (stream: MediaStream) => {
//   return new AudioRecorder(stream) satisfies MediaRecorder;
// };

/**
 * A MediaRecord implementation only meant for recording audio streams.
 *
 * @private
 */
class MediaRecorder extends _reactNativeWebrtc.EventTarget {
  mimeType = 'audio/pcm';
  audioBitsPerSecond = 0; // TODO?
  state = 'inactive';
  videoBitsPerSecond = 0; // TODO?
  audioBitrateMode = 'constant';
  _reactTag = undefined;
  _parts = [];
  constructor(stream) {
    super();
    this.stream = stream;
  }
  registerListener() {
    let audioTracks = this.stream.getAudioTracks();
    if (audioTracks.length !== 1) {
      return;
    }
    const mediaStreamTrack = audioTracks[0];
    const peerConnectionId = mediaStreamTrack._peerConnectionId ?? -1;
    const mediaStreamTrackId = mediaStreamTrack?.id;
    this._reactTag = _LKNativeModule.default.createAudioSinkListener(peerConnectionId, mediaStreamTrackId);
    (0, _EventEmitter.addListener)(this, 'LK_AUDIO_DATA', event => {
      if (this._reactTag && event.id === this._reactTag && this.state === 'recording') {
        let str = event.data;
        this._parts.push(str);
      }
    });
  }
  unregisterListener() {
    if (this._reactTag) {
      let audioTracks = this.stream.getAudioTracks();
      if (audioTracks.length !== 1) {
        _logger.log.error("couldn't find any audio tracks to record from!");
        return;
      }
      const mediaStreamTrack = audioTracks[0];
      const peerConnectionId = mediaStreamTrack._peerConnectionId ?? -1;
      const mediaStreamTrackId = mediaStreamTrack?.id;
      _LKNativeModule.default.deleteAudioSinkListener(this._reactTag, peerConnectionId, mediaStreamTrackId);
    }
  }
  pause() {
    this.state = 'paused';
    this.dispatchEvent(new _reactNativeWebrtc.Event('pause'));
  }
  resume() {
    this.state = 'recording';
    this.dispatchEvent(new _reactNativeWebrtc.Event('resume'));
  }
  start() {
    this.registerListener();
    this.state = 'recording';
    this.dispatchEvent(new _reactNativeWebrtc.Event('start'));
  }
  stop() {
    // dispatch data must come before stopping.
    this.dispatchData();
    this.unregisterListener();
    this.state = 'inactive';
    this.dispatchEvent(new _reactNativeWebrtc.Event('stop'));
  }
  requestData() {
    this.dispatchData();
  }
  dispatchData() {
    let combinedStr = this._parts.reduce((sum, cur) => sum + cur, '');
    let data = (0, _base64Js.toByteArray)(combinedStr);
    this._parts = [];
    this.dispatchEvent(new BlobEvent('dataavailable', {
      data: {
        byteArray: data
      }
    }));
  }
  get ondataavailable() {
    return (0, _reactNativeWebrtc.getEventAttributeValue)(this, 'dataavailable');
  }
  set ondataavailable(value) {
    (0, _reactNativeWebrtc.setEventAttributeValue)(this, 'dataavailable', value);
  }
  get onerror() {
    return (0, _reactNativeWebrtc.getEventAttributeValue)(this, 'error');
  }
  set onerror(value) {
    (0, _reactNativeWebrtc.setEventAttributeValue)(this, 'error', value);
  }
  get onpause() {
    return (0, _reactNativeWebrtc.getEventAttributeValue)(this, 'pause');
  }
  set onpause(value) {
    (0, _reactNativeWebrtc.setEventAttributeValue)(this, 'pause', value);
  }
  get onresume() {
    return (0, _reactNativeWebrtc.getEventAttributeValue)(this, 'resume');
  }
  set onresume(value) {
    (0, _reactNativeWebrtc.setEventAttributeValue)(this, 'resume', value);
  }
  get onstart() {
    return (0, _reactNativeWebrtc.getEventAttributeValue)(this, 'start');
  }
  set onstart(value) {
    (0, _reactNativeWebrtc.setEventAttributeValue)(this, 'start', value);
  }
  get onstop() {
    return (0, _reactNativeWebrtc.getEventAttributeValue)(this, 'stop');
  }
  set onstop(value) {
    (0, _reactNativeWebrtc.setEventAttributeValue)(this, 'stop', value);
  }
}

/**
 * @eventClass
 * This event is fired whenever the Track is changed in PeerConnection.
 * @param {TRACK_EVENTS} type - The type of event.
 * @param {IRTCTrackEventInitDict} eventInitDict - The event init properties.
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/track_event MDN} for details.
 */
exports.MediaRecorder = MediaRecorder;
class BlobEvent extends _reactNativeWebrtc.Event {
  /** @eventProperty */

  constructor(type, eventInitDict) {
    super(type, eventInitDict);
    this.data = eventInitDict.data;
  }
}
