// @livekit/react-native/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  registerGlobals: true,
  AudioDeviceModule: true,
  AudioEngineMuteMode: true,
  AudioEngineAvailability: true,
  audioDeviceModuleEvents: true,
  AudioSession: true,
  AndroidAudioTypePresets: true,
  RNE2EEManager: true,
  RNKeyProvider: true
};
Object.defineProperty(exports, "AndroidAudioTypePresets", {
  enumerable: true,
  get: function () {
    return _AudioSession.AndroidAudioTypePresets;
  }
});
Object.defineProperty(exports, "AudioDeviceModule", {
  enumerable: true,
  get: function () {
    return _reactNativeWebrtc.AudioDeviceModule;
  }
});
Object.defineProperty(exports, "AudioEngineAvailability", {
  enumerable: true,
  get: function () {
    return _reactNativeWebrtc.AudioEngineAvailability;
  }
});
Object.defineProperty(exports, "AudioEngineMuteMode", {
  enumerable: true,
  get: function () {
    return _reactNativeWebrtc.AudioEngineMuteMode;
  }
});
Object.defineProperty(exports, "AudioSession", {
  enumerable: true,
  get: function () {
    return _AudioSession.default;
  }
});
Object.defineProperty(exports, "RNE2EEManager", {
  enumerable: true,
  get: function () {
    return _RNE2EEManager.default;
  }
});
Object.defineProperty(exports, "RNKeyProvider", {
  enumerable: true,
  get: function () {
    return _RNKeyProvider.default;
  }
});
Object.defineProperty(exports, "audioDeviceModuleEvents", {
  enumerable: true,
  get: function () {
    return _reactNativeWebrtc.audioDeviceModuleEvents;
  }
});
exports.registerGlobals = registerGlobals;
require("well-known-symbols/Symbol.asyncIterator/auto");
require("well-known-symbols/Symbol.iterator/auto");
require("./polyfills/DOMException");
require("./polyfills/MediaRecorderShim");
var _reactNativeWebrtc = require("@livekit/react-native-webrtc");
var _reactNativeUrlPolyfill = require("react-native-url-polyfill");
require("./polyfills/EncoderDecoderTogether.min.js");
var _AudioSession = _interopRequireWildcard(require("./audio/AudioSession"));
var _reactNative = require("react-native");
var _AudioManager = require("./audio/AudioManager");
Object.keys(_AudioManager).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _AudioManager[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _AudioManager[key];
    }
  });
});
var _RNE2EEManager = _interopRequireDefault(require("./e2ee/RNE2EEManager"));
var _RNKeyProvider = _interopRequireDefault(require("./e2ee/RNKeyProvider"));
var _EventEmitter = require("./events/EventEmitter");
var _webStreamsPolyfill = require("web-streams-polyfill");
var _hooks = require("./hooks");
Object.keys(_hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _hooks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _hooks[key];
    }
  });
});
var _BarVisualizer = require("./components/BarVisualizer");
Object.keys(_BarVisualizer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _BarVisualizer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _BarVisualizer[key];
    }
  });
});
var _LiveKitRoom = require("./components/LiveKitRoom");
Object.keys(_LiveKitRoom).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _LiveKitRoom[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _LiveKitRoom[key];
    }
  });
});
var _VideoTrack = require("./components/VideoTrack");
Object.keys(_VideoTrack).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _VideoTrack[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _VideoTrack[key];
    }
  });
});
var _VideoView = require("./components/VideoView");
Object.keys(_VideoView).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _VideoView[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _VideoView[key];
    }
  });
});
var _useParticipant = require("./useParticipant");
Object.keys(_useParticipant).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useParticipant[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useParticipant[key];
    }
  });
});
var _useRoom = require("./useRoom");
Object.keys(_useRoom).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useRoom[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useRoom[key];
    }
  });
});
var _logger = require("./logger");
Object.keys(_logger).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _logger[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _logger[key];
    }
  });
});
var _AudioManagerLegacy = require("./audio/AudioManagerLegacy");
Object.keys(_AudioManagerLegacy).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _AudioManagerLegacy[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _AudioManagerLegacy[key];
    }
  });
});
var _MediaRecorder = require("./audio/MediaRecorder");
Object.keys(_MediaRecorder).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _MediaRecorder[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _MediaRecorder[key];
    }
  });
});
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
// Automatic polyfills, keep these at the top.

// Caution: This has a transitive import of livekit-client, keep last.

/**
 * Registers the required globals needed for LiveKit to work.
 *
 * Must be called before using LiveKit.
 *
 * @param options Optional configuration for global registration
 */
function registerGlobals(options) {
  const opts = {
    autoConfigureAudioSession: true,
    ...options
  };
  (0, _reactNativeWebrtc.registerGlobals)();
  if (opts.autoConfigureAudioSession) {
    (0, _AudioManager.setupIOSAudioManagement)();
  }
  livekitRegisterGlobals();
  (0, _reactNativeUrlPolyfill.setupURLPolyfill)();
  fixWebrtcAdapter();
  shimPromiseAllSettled();
  shimArrayAt();
  shimCryptoUuid();
  shimWebstreams();
  (0, _EventEmitter.setupNativeEvents)();
}
function livekitRegisterGlobals() {
  let lkGlobal = {
    platform: _reactNative.Platform.OS,
    devicePixelRatio: _reactNative.PixelRatio.get()
  };

  // @ts-ignore
  global.LiveKitReactNativeGlobal = lkGlobal;
}
function fixWebrtcAdapter() {
  // @ts-ignore
  if (window?.navigator !== undefined) {
    // @ts-ignore
    const {
      navigator
    } = window;
    if (navigator.userAgent === undefined) {
      navigator.userAgent = navigator.product ?? 'Unknown';
    }
  }
}
function shimPromiseAllSettled() {
  var allSettled = require('promise.allsettled');
  allSettled.shim();
}
function shimArrayAt() {
  // Some versions of RN don't have Array.prototype.at, which is used by sdp-transform
  if (!Array.prototype.at) {
    var at = require('array.prototype.at');
    at.shim();
  }
}
function shimCryptoUuid() {
  let crypto = global.crypto;
  if (typeof global.crypto?.randomUUID !== 'function') {
    let createRandomUUID = () => {
      return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
        /* eslint-disable no-bitwise */
        const r = Math.random() * 16 | 0;
        const v = c === 'x' ? r : r & 0x3 | 0x8;
        return v.toString(16);
      });
    };
    if (!crypto) {
      crypto = {};
      global.crypto = crypto;
    }
    crypto.randomUUID = createRandomUUID;
  }
}
function shimWebstreams() {
  // @ts-expect-error: global.WritableStream isn't typed here.
  if (typeof global.WritableStream === 'undefined') {
    // @ts-expect-error
    global.WritableStream = _webStreamsPolyfill.WritableStream;
  }

  // @ts-expect-error: global.ReadableStream isn't typed here.
  if (typeof global.ReadableStream === 'undefined') {
    // @ts-expect-error
    global.ReadableStream = _webStreamsPolyfill.ReadableStream;
  }
  // @ts-expect-error: global.CountQueuingStrategy isn't typed here.
  if (typeof global.CountQueuingStrategy === 'undefined') {
    // @ts-expect-error
    global.CountQueuingStrategy = _webStreamsPolyfill.CountQueuingStrategy;
  }
  // @ts-expect-error: global.TransformStream isn't typed here.
  if (typeof global.TransformStream === 'undefined') {
    // @ts-expect-error
    global.TransformStream = _webStreamsPolyfill.TransformStream;
  }
}

// deprecated
// deprecated
// deprecated
