// @livekit/react-native/lib/module/components/VideoTrack.js
"use strict";

import { Platform, StyleSheet, View } from 'react-native';
import { LocalVideoTrack, TrackEvent } from 'livekit-client';
import { RTCView, RTCPIPView } from '@livekit/react-native-webrtc';
import { forwardRef, useCallback, useEffect, useMemo, useState } from 'react';
import { RemoteVideoTrack } from 'livekit-client';
import ViewPortDetector from './ViewPortDetector';

/**
 * Props for the VideoTrack component.
 * @public
 */

/**
 * @inline
 */
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * VideoTrack component for displaying video tracks in a React Native application.
 * It supports both local and remote video tracks from LiveKit, and handles adaptive streaming for remote tracks.
 *
 * @param props - See VideoTrackProps for details.
 * @returns A React component that renders the given video track.
 * @public
 */
export const VideoTrack = /*#__PURE__*/forwardRef(({
  style = {},
  trackRef,
  objectFit = 'cover',
  zOrder,
  mirror,
  iosPIP
}, ref) => {
  const [elementInfo] = useState(() => {
    let info = new VideoTrackElementInfo();
    info.id = trackRef?.publication?.trackSid;
    return info;
  });
  const layoutOnChange = useCallback(event => elementInfo.onLayout(event), [elementInfo]);
  const iosPIPEnabled = iosPIP?.enabled ?? false;
  const visibilityOnChange = useCallback(isVisible => elementInfo.onVisibility(isVisible || iosPIPEnabled), [elementInfo, iosPIPEnabled]);
  const videoTrack = trackRef?.publication.track;
  const shouldObserveVisibility = useMemo(() => {
    return videoTrack instanceof RemoteVideoTrack && videoTrack.isAdaptiveStream;
  }, [videoTrack]);
  const [mediaStream, setMediaStream] = useState(videoTrack?.mediaStream);
  useEffect(() => {
    setMediaStream(videoTrack?.mediaStream);
    if (videoTrack instanceof LocalVideoTrack) {
      const onRestarted = track => {
        setMediaStream(track?.mediaStream);
      };
      videoTrack.on(TrackEvent.Restarted, onRestarted);
      return () => {
        videoTrack.off(TrackEvent.Restarted, onRestarted);
      };
    } else {
      return () => {};
    }
  }, [videoTrack]);
  useEffect(() => {
    if (videoTrack instanceof RemoteVideoTrack && videoTrack.isAdaptiveStream) {
      videoTrack?.observeElementInfo(elementInfo);
      return () => {
        videoTrack?.stopObservingElementInfo(elementInfo);
      };
    } else {
      return () => {};
    }
  }, [videoTrack, elementInfo]);
  let videoView;
  if (!iosPIP || Platform.OS !== 'ios') {
    videoView = /*#__PURE__*/_jsx(RTCView, {
      style: styles.videoTrack
      // @ts-expect-error
      ,
      streamURL: mediaStream?.toURL() ?? '',
      objectFit: objectFit,
      zOrder: zOrder,
      mirror: mirror
      // @ts-ignore
      ,
      ref: ref
    });
  } else {
    videoView = /*#__PURE__*/_jsx(RTCPIPView, {
      style: styles.videoTrack
      // @ts-expect-error
      ,
      streamURL: mediaStream?.toURL() ?? '',
      objectFit: objectFit,
      zOrder: zOrder,
      mirror: mirror
      // TODO: fix this up in react-native-webrtc side.
      // @ts-expect-error
      ,
      iosPIP: iosPIP,
      ref: ref
    });
  }
  return /*#__PURE__*/_jsx(View, {
    style: {
      ...style,
      ...styles.container
    },
    onLayout: layoutOnChange,
    children: /*#__PURE__*/_jsx(ViewPortDetector, {
      onChange: visibilityOnChange,
      style: styles.videoTrack,
      disabled: !shouldObserveVisibility,
      propKey: videoTrack,
      children: videoView
    })
  });
});
const styles = StyleSheet.create({
  container: {},
  videoTrack: {
    flex: 1,
    width: '100%'
  }
});
class VideoTrackElementInfo {
  element = {};
  _width = 0;
  _height = 0;
  _observing = false;
  visible = true;
  pictureInPicture = false;
  width = () => this._width;
  height = () => this._height;
  observe() {
    this._observing = true;
  }
  stopObserving() {
    this._observing = false;
  }
  onLayout(event) {
    let {
      width,
      height
    } = event.nativeEvent.layout;
    this._width = width;
    this._height = height;
    if (this._observing) {
      this.handleResize?.();
    }
  }
  onVisibility(isVisible) {
    if (this.visible !== isVisible) {
      this.visible = isVisible;
      this.visibilityChangedAt = Date.now();
      if (this._observing) {
        this.handleVisibilityChanged?.();
      }
    }
  }
}
