// @livekit/react-native/lib/module/components/VideoView.js
"use strict";

import { StyleSheet, View } from 'react-native';
import { LocalVideoTrack, TrackEvent } from 'livekit-client';
import { RTCView } from '@livekit/react-native-webrtc';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { RemoteVideoTrack } from 'livekit-client';
import ViewPortDetector from './ViewPortDetector';

/**
 * @deprecated use `VideoTrack` and `VideoTrackProps` instead.
 */
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * @deprecated use `VideoTrack` and `VideoTrackProps` instead.
 */
export const VideoView = ({
  style = {},
  videoTrack,
  objectFit = 'cover',
  zOrder,
  mirror
}) => {
  const [elementInfo] = useState(() => {
    let info = new VideoViewElementInfo();
    info.id = videoTrack?.sid;
    info.something = videoTrack;
    return info;
  });
  const layoutOnChange = useCallback(event => elementInfo.onLayout(event), [elementInfo]);
  const visibilityOnChange = useCallback(isVisible => elementInfo.onVisibility(isVisible), [elementInfo]);
  const shouldObserveVisibility = useMemo(() => {
    return videoTrack instanceof RemoteVideoTrack && videoTrack.isAdaptiveStream;
  }, [videoTrack]);
  const [mediaStream, setMediaStream] = useState(videoTrack?.mediaStream);
  useEffect(() => {
    setMediaStream(videoTrack?.mediaStream);
    if (videoTrack instanceof LocalVideoTrack) {
      const onRestarted = track => {
        setMediaStream(track?.mediaStream);
      };
      videoTrack.on(TrackEvent.Restarted, onRestarted);
      return () => {
        videoTrack.off(TrackEvent.Restarted, onRestarted);
      };
    } else {
      return () => {};
    }
  }, [videoTrack]);
  useEffect(() => {
    if (videoTrack instanceof RemoteVideoTrack && videoTrack.isAdaptiveStream) {
      videoTrack?.observeElementInfo(elementInfo);
      return () => {
        videoTrack?.stopObservingElementInfo(elementInfo);
      };
    } else {
      return () => {};
    }
  }, [videoTrack, elementInfo]);
  return /*#__PURE__*/_jsx(View, {
    style: {
      ...style,
      ...styles.container
    },
    onLayout: layoutOnChange,
    children: /*#__PURE__*/_jsx(ViewPortDetector, {
      onChange: visibilityOnChange,
      style: styles.videoView,
      disabled: !shouldObserveVisibility,
      propKey: videoTrack,
      children: /*#__PURE__*/_jsx(RTCView, {
        style: styles.videoView
        // @ts-expect-error
        ,
        streamURL: mediaStream?.toURL() ?? '',
        objectFit: objectFit,
        zOrder: zOrder,
        mirror: mirror
      })
    })
  });
};
const styles = StyleSheet.create({
  container: {},
  videoView: {
    flex: 1,
    width: '100%'
  }
});
class VideoViewElementInfo {
  element = {};
  _width = 0;
  _height = 0;
  _observing = false;
  visible = true;
  pictureInPicture = false;
  width = () => this._width;
  height = () => this._height;
  observe() {
    this._observing = true;
  }
  stopObserving() {
    this._observing = false;
  }
  onLayout(event) {
    let {
      width,
      height
    } = event.nativeEvent.layout;
    this._width = width;
    this._height = height;
    if (this._observing) {
      this.handleResize?.();
    }
  }
  onVisibility(isVisible) {
    if (this.visible !== isVisible) {
      this.visible = isVisible;
      this.visibilityChangedAt = Date.now();
      if (this._observing) {
        this.handleVisibilityChanged?.();
      }
    }
  }
}
