// @livekit/react-native/lib/module/components/ViewPortDetector.js
'use strict';

import { Component } from 'react';
import { AppState, View } from 'react-native';
import { jsx as _jsx } from "react/jsx-runtime";
const DEFAULT_DELAY = 1000;
class TimeoutHandler {
  handlerRef = {
    id: -1
  };
  get handler() {
    return this.handlerRef.id;
  }
  set handler(n) {
    this.handlerRef.id = n;
  }
  clear() {
    clearTimeout(this.handlerRef.id);
  }
}
function setIntervalWithTimeout(callback, intervalMs, handleWrapper = new TimeoutHandler()) {
  let cleared = false;
  const timeout = () => {
    handleWrapper.handler = setTimeout(() => {
      callback(() => {
        cleared = true;
        handleWrapper.clear();
      });
      if (!cleared) {
        timeout();
      }
    }, intervalMs);
  };
  timeout();
  return handleWrapper;
}

/**
 * Detects when this is in the viewport and visible.
 *
 * Will not fire visibility changes for zero width/height components.
 */
export default class ViewPortDetector extends Component {
  lastValue = null;
  interval = null;
  view = null;
  lastAppStateActive = false;
  appStateSubscription = null;
  constructor(props) {
    super(props);
    this.state = {
      rectTop: 0,
      rectBottom: 0
    };
  }
  componentDidMount() {
    this.lastAppStateActive = AppState.currentState === 'active';
    this.appStateSubscription = AppState.addEventListener('change', this.handleAppStateChange);
    if (this.hasValidTimeout(this.props.disabled, this.props.delay)) {
      this.startWatching();
    }
  }
  componentWillUnmount() {
    this.appStateSubscription?.remove();
    this.appStateSubscription = null;
    this.stopWatching();
  }
  hasValidTimeout = (disabled, delay) => {
    let disabledValue = disabled ?? false;
    let delayValue = delay ?? DEFAULT_DELAY;
    return AppState.currentState === 'active' && !disabledValue && delayValue > 0;
  };
  UNSAFE_componentWillReceiveProps(nextProps) {
    if (!this.hasValidTimeout(nextProps.disabled, nextProps.delay)) {
      this.stopWatching();
    } else {
      if (this.props.propKey !== nextProps.propKey) {
        this.lastValue = null;
      }
      this.startWatching();
    }
  }
  handleAppStateChange = nextAppState => {
    let nextAppStateActive = nextAppState === 'active';
    if (this.lastAppStateActive !== nextAppStateActive) {
      this.checkVisibility();
    }
    this.lastAppStateActive = nextAppStateActive;
    if (!this.hasValidTimeout(this.props.disabled, this.props.delay)) {
      this.stopWatching();
    } else {
      this.startWatching();
    }
  };
  startWatching = () => {
    if (this.interval) {
      return;
    }
    this.interval = setIntervalWithTimeout(this.checkVisibility, this.props.delay || DEFAULT_DELAY);
  };
  stopWatching = () => {
    this.interval?.clear();
    this.interval = null;
  };
  checkVisibility = () => {
    if (!this.view) {
      return;
    }
    if (AppState.currentState !== 'active') {
      this.updateVisibility(false);
      return;
    }
    this.view.measure((_x, _y, width, height, _pageX, _pageY) => {
      this.checkInViewPort(width, height);
    });
  };
  checkInViewPort = (width, height) => {
    let isVisible;
    // Not visible if any of these are missing.
    if (!width || !height) {
      isVisible = false;
    } else {
      isVisible = true;
    }
    this.updateVisibility(isVisible);
  };
  updateVisibility = isVisible => {
    if (this.lastValue !== isVisible) {
      this.lastValue = isVisible;
      this.props.onChange?.(isVisible);
    }
  };
  render() {
    return /*#__PURE__*/_jsx(View, {
      collapsable: false,
      ref: component => {
        this.view = component;
      },
      ...this.props,
      children: this.props.children
    });
  }
}
