// @livekit/react-native-webrtc/lib/module/AudioDeviceModuleEvents.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
import { NativeModules, Platform } from 'react-native';
import { addListener, removeListener } from './EventEmitter';
const {
  WebRTCModule
} = NativeModules;

// Empty object for events with no data

/**
 * Handler function that must return a number (0 for success, non-zero for error)
 */

/**
 * Event emitter for RTCAudioDeviceModule delegate callbacks.
 * iOS/macOS only.
 */
class AudioDeviceModuleEventEmitter {
  constructor() {
    _defineProperty(this, "engineCreatedHandler", null);
    _defineProperty(this, "willEnableEngineHandler", null);
    _defineProperty(this, "willStartEngineHandler", null);
    _defineProperty(this, "didStopEngineHandler", null);
    _defineProperty(this, "didDisableEngineHandler", null);
    _defineProperty(this, "willReleaseEngineHandler", null);
  }
  setupListeners() {
    if (Platform.OS !== 'android' && WebRTCModule) {
      // Setup handlers for blocking delegate methods
      addListener(this, 'audioDeviceModuleEngineCreated', async () => {
        let result = 0;
        if (this.engineCreatedHandler) {
          try {
            await this.engineCreatedHandler();
          } catch (error) {
            // If error is a number, use it as the error code, otherwise use -1
            result = typeof error === 'number' ? error : -1;
          }
        }
        WebRTCModule.audioDeviceModuleResolveEngineCreated(result);
      });
      addListener(this, 'audioDeviceModuleEngineWillEnable', async event => {
        const {
          isPlayoutEnabled,
          isRecordingEnabled
        } = event;
        let result = 0;
        if (this.willEnableEngineHandler) {
          try {
            await this.willEnableEngineHandler({
              isPlayoutEnabled,
              isRecordingEnabled
            });
          } catch (error) {
            // If error is a number, use it as the error code, otherwise use -1
            result = typeof error === 'number' ? error : -1;
          }
        }
        WebRTCModule.audioDeviceModuleResolveWillEnableEngine(result);
      });
      addListener(this, 'audioDeviceModuleEngineWillStart', async event => {
        const {
          isPlayoutEnabled,
          isRecordingEnabled
        } = event;
        let result = 0;
        if (this.willStartEngineHandler) {
          try {
            await this.willStartEngineHandler({
              isPlayoutEnabled,
              isRecordingEnabled
            });
          } catch (error) {
            // If error is a number, use it as the error code, otherwise use -1
            result = typeof error === 'number' ? error : -1;
          }
        }
        WebRTCModule.audioDeviceModuleResolveWillStartEngine(result);
      });
      addListener(this, 'audioDeviceModuleEngineDidStop', async event => {
        const {
          isPlayoutEnabled,
          isRecordingEnabled
        } = event;
        let result = 0;
        if (this.didStopEngineHandler) {
          try {
            await this.didStopEngineHandler({
              isPlayoutEnabled,
              isRecordingEnabled
            });
          } catch (error) {
            // If error is a number, use it as the error code, otherwise use -1
            result = typeof error === 'number' ? error : -1;
          }
        }
        WebRTCModule.audioDeviceModuleResolveDidStopEngine(result);
      });
      addListener(this, 'audioDeviceModuleEngineDidDisable', async event => {
        const {
          isPlayoutEnabled,
          isRecordingEnabled
        } = event;
        let result = 0;
        if (this.didDisableEngineHandler) {
          try {
            await this.didDisableEngineHandler({
              isPlayoutEnabled,
              isRecordingEnabled
            });
          } catch (error) {
            // If error is a number, use it as the error code, otherwise use -1
            result = typeof error === 'number' ? error : -1;
          }
        }
        WebRTCModule.audioDeviceModuleResolveDidDisableEngine(result);
      });
      addListener(this, 'audioDeviceModuleEngineWillRelease', async () => {
        let result = 0;
        if (this.willReleaseEngineHandler) {
          try {
            await this.willReleaseEngineHandler();
          } catch (error) {
            // If error is a number, use it as the error code, otherwise use -1
            result = typeof error === 'number' ? error : -1;
          }
        }
        WebRTCModule.audioDeviceModuleResolveWillReleaseEngine(result);
      });
    }
  }

  /**
   * Subscribe to speech activity events (started/ended)
   */
  addSpeechActivityListener(listener) {
    addListener(listener, 'audioDeviceModuleSpeechActivity', listener);
  }

  /**
   * Remove a previously registered speech activity listener
   */
  removeSpeechActivityListener(listener) {
    removeListener(listener);
  }

  /**
   * Subscribe to devices updated event (input/output devices changed)
   */
  addDevicesUpdatedListener(listener) {
    addListener(listener, 'audioDeviceModuleDevicesUpdated', listener);
  }

  /**
   * Remove a previously registered devices updated listener
   */
  removeDevicesUpdatedListener(listener) {
    removeListener(listener);
  }

  /**
   * Set handler for engine created delegate - MUST return 0 for success or error code
   * This handler blocks the native thread until it returns, throw to cancel audio engine's operation
   */
  setEngineCreatedHandler(handler) {
    this.engineCreatedHandler = handler;
  }

  /**
   * Set handler for will enable engine delegate - MUST return 0 for success or error code
   * This handler blocks the native thread until it returns, throw to cancel audio engine's operation
   */
  setWillEnableEngineHandler(handler) {
    this.willEnableEngineHandler = handler;
  }

  /**
   * Set handler for will start engine delegate - MUST return 0 for success or error code
   * This handler blocks the native thread until it returns, throw to cancel audio engine's operation
   */
  setWillStartEngineHandler(handler) {
    this.willStartEngineHandler = handler;
  }

  /**
   * Set handler for did stop engine delegate - MUST return 0 for success or error code
   * This handler blocks the native thread until it returns, throw to cancel audio engine's operation
   */
  setDidStopEngineHandler(handler) {
    this.didStopEngineHandler = handler;
  }

  /**
   * Set handler for did disable engine delegate - MUST return 0 for success or error code
   * This handler blocks the native thread until it returns, throw to cancel audio engine's operation
   */
  setDidDisableEngineHandler(handler) {
    this.didDisableEngineHandler = handler;
  }

  /**
   * Set handler for will release engine delegate
   * This handler blocks the native thread until it returns, throw to cancel audio engine's operation
   */
  setWillReleaseEngineHandler(handler) {
    this.willReleaseEngineHandler = handler;
  }
}
export const audioDeviceModuleEvents = new AudioDeviceModuleEventEmitter();
