// @livekit/react-native/lib/module/e2ee/RNE2EEManager.js
"use strict";

import { RTCDataPacketCryptorFactory, RTCFrameCryptorAlgorithm, RTCFrameCryptorFactory } from '@livekit/react-native-webrtc';
import { ParticipantEvent, RoomEvent, EncryptionEvent, Mutex } from 'livekit-client';
import EventEmitter from 'events';
/**
 * @experimental
 */
export default class RNE2EEManager extends EventEmitter {
  frameCryptors = new Map();
  algorithm = RTCFrameCryptorAlgorithm.kAesGcm;
  encryptionEnabled = false;
  dataChannelEncryptionEnabled = false;
  dataPacketCryptorLock = new Mutex();
  dataPacketCryptor = undefined;
  constructor(keyProvider, dcEncryptionEnabled = false) {
    super();
    this.keyProvider = keyProvider;
    this.encryptionEnabled = false;
    this.dataChannelEncryptionEnabled = dcEncryptionEnabled;
  }
  get isEnabled() {
    return this.encryptionEnabled;
  }
  get isDataChannelEncryptionEnabled() {
    return this.isEnabled && this.dataChannelEncryptionEnabled;
  }
  set isDataChannelEncryptionEnabled(value) {
    this.dataChannelEncryptionEnabled = value;
  }
  setup(room) {
    if (this.room !== room) {
      this.room = room;
      this.setupEventListeners(room);
    }
  }
  setupEventListeners(room) {
    room.localParticipant.on(ParticipantEvent.LocalTrackPublished, async publication => {
      this.setupE2EESender(publication, room.localParticipant);
    }).on(ParticipantEvent.LocalTrackUnpublished, async publication => {
      let frameCryptor = this.findTrackCryptor(publication.trackSid);
      if (frameCryptor) {
        this.frameCryptors.delete(publication.trackSid);
        await frameCryptor.setEnabled(false);
        await frameCryptor.dispose();
      }
    });
    room.on(RoomEvent.TrackSubscribed, (_track, pub, participant) => {
      this.setupE2EEReceiver(pub, participant);
    }).on(RoomEvent.TrackUnsubscribed, async (_track, publication, _participant) => {
      let frameCryptor = this.findTrackCryptor(publication.trackSid);
      if (frameCryptor) {
        this.frameCryptors.delete(publication.trackSid);
        await frameCryptor.setEnabled(false);
        await frameCryptor.dispose();
      }
    }).on(RoomEvent.SignalConnected, () => {
      if (!this.room) {
        throw new TypeError(`expected room to be present on signal connect`);
      }
      this.setParticipantCryptorEnabled(this.room.localParticipant.isE2EEEnabled, this.room.localParticipant.identity);
    });
  }
  async setupE2EESender(publication, participant) {
    if (!publication.isEncrypted) {
      return;
    }
    var frameCryptor = this.findTrackCryptor(publication.trackSid);
    if (!frameCryptor) {
      frameCryptor = this.createFrameCryptorForSender(publication.track.sender, participant.identity);
      this.frameCryptors.set(publication.trackSid, frameCryptor);
      frameCryptor.setEnabled(true);
      frameCryptor.setKeyIndex(this.keyProvider.getLatestKeyIndex(participant.identity));
    }
  }
  async setupE2EEReceiver(publication, participant) {
    if (!publication.isEncrypted) {
      return;
    }
    var frameCryptor = this.findTrackCryptor(publication.trackSid);
    if (!frameCryptor) {
      frameCryptor = this.createFrameCryptorForReceiver(publication.track.receiver, participant.identity);
      this.frameCryptors.set(publication.trackSid, frameCryptor);
      frameCryptor.setEnabled(true);
      frameCryptor.setKeyIndex(this.keyProvider.getLatestKeyIndex(participant.identity));
    }
  }
  setSifTrailer(trailer) {
    this.keyProvider.setSifTrailer(trailer);
  }
  async getDataPacketCryptor() {
    let dataPacketCryptor = this.dataPacketCryptor;
    if (dataPacketCryptor) {
      return dataPacketCryptor;
    }
    let unlock = await this.dataPacketCryptorLock.lock();
    try {
      dataPacketCryptor = this.dataPacketCryptor;
      if (dataPacketCryptor) {
        return dataPacketCryptor;
      }
      dataPacketCryptor = await RTCDataPacketCryptorFactory.createDataPacketCryptor(this.algorithm, this.keyProvider.rtcKeyProvider);
      this.dataPacketCryptor = dataPacketCryptor;
      return dataPacketCryptor;
    } finally {
      unlock();
    }
  }
  async encryptData(data) {
    let room = this.room;
    if (!room) {
      throw new Error("e2eemanager isn't setup with room!");
    }
    let participantId = room.localParticipant.identity;
    let dataPacketCryptor = await this.getDataPacketCryptor();
    let encryptedPacket = await dataPacketCryptor.encrypt(participantId, this.keyProvider.getLatestKeyIndex(participantId), data);
    if (!encryptedPacket) {
      throw new Error('encryption for packet failed');
    }
    return {
      uuid: '',
      //not used
      payload: encryptedPacket.payload,
      iv: encryptedPacket.iv,
      keyIndex: encryptedPacket.keyIndex
    };
  }
  async handleEncryptedData(payload, iv, participantIdentity, keyIndex) {
    let packet = {
      payload,
      iv,
      keyIndex
    };
    let dataPacketCryptor = await this.getDataPacketCryptor();
    let decryptedData = await dataPacketCryptor.decrypt(participantIdentity, packet);
    if (!decryptedData) {
      throw new Error('decryption for packet failed');
    }
    return {
      uuid: '',
      //not used
      payload: decryptedData
    };
  }

  // Utility methods
  //////////////////////

  findTrackCryptor(trackId) {
    return this.frameCryptors.get(trackId);
  }
  createFrameCryptorForSender(sender, participantId) {
    return RTCFrameCryptorFactory.createFrameCryptorForRtpSender(participantId, sender, this.algorithm, this.keyProvider.rtcKeyProvider);
  }
  createFrameCryptorForReceiver(receiver, participantId) {
    return RTCFrameCryptorFactory.createFrameCryptorForRtpReceiver(participantId, receiver, this.algorithm, this.keyProvider.rtcKeyProvider);
  }
  setupEngine(_engine) {
    // No-op
  }
  setParticipantCryptorEnabled(enabled, participantIdentity) {
    if (this.encryptionEnabled !== enabled && participantIdentity === this.room?.localParticipant.identity) {
      this.encryptionEnabled = enabled;
      this.emit(EncryptionEvent.ParticipantEncryptionStatusChanged, enabled, this.room.localParticipant);
    } else {
      const participant = this.room?.getParticipantByIdentity(participantIdentity);
      if (!participant) {
        throw TypeError(`couldn't set encryption status, participant not found ${participantIdentity}`);
      }
      this.emit(EncryptionEvent.ParticipantEncryptionStatusChanged, enabled, participant);
    }
  }
}
